parser/src/parser.rs

```diff
@@ -0,0 +1,523 @@
+#![allow(clippy::type_complexity)]
+
+use nom::{
+    branch::alt,
+    combinator::{map, map_res},
+    error::{Error as NomError, ErrorKind, ParseError},
+    multi::{fold_many0, many0, separated_list0},
+    sequence::{pair, tuple},
+    IResult,
+};
+
+use crate::{
+    ast::{
+        ArithmBinaryOp, Ast, BasicBlock, BinaryOp, BitBinaryOp, Call, Function, GotoInsn, Insn,
+        Literal, LogicalBinaryOp, RValue, Type, UnaryOp, Value, Variable,
+    },
+    lexer::{
+        Keyword, Lexer,
+        Token::{self, *},
+    },
+};
+
+use super::{Error, Result};
+
+/// Combinator that takes one token from the input.
+fn take1(i: &[Token]) -> IResult<&[Token], Token> {
+    let first_token = i
+        .first()
+        .ok_or_else(|| nom::Err::Error(NomError::from_error_kind(i, ErrorKind::Complete)))?;
+    Ok((&i[1..], first_token.clone()))
```

Comment:
Can we not to clone, but just move `first_token`?
```suggestion
    Ok((&i[1..], first_token.to_owned()))
```
Or is it bad if we received the `i` array by reference and moved the first element from it?
Maybe we'll change the signature of our parser to `Fn(Vec<Token>) -> IResult<Vec<Token>, Token>`? It seems that after the parser has accepted the array, no one now needs an array with the first element.

Reply:
I also suggest a small refactoring using `.split_first()`:
```diff
-    let first_token = i
-        .first()
-        .ok_or_else(|| nom::Err::Error(NomError::from_error_kind(i, ErrorKind::Complete)))?;
-    Ok((&i[1..], first_token.clone()))
+    let (first_token, remaining) = i
+        .split_first()
+        .ok_or_else(|| nom_error(i, ErrorKind::Complete))?;
+    Ok((remaining, first_token.to_owned()))
```

Reply:
1. `to_owned()` clones under the hood. If you got a reference, you have no way to move from it, because you are not an owner of this memory, you just "borrowed" it. It is a common rule that always true.
`ToOwned` is just a `Clone` generalization that allows to use custom owning type.

`ToOwned` trivially implemented for all types that implement clone:
```rust
impl<T> ToOwned for T
where
    T: Clone,
{
    type Owned = T;
    fn to_owned(&self) -> T {
        self.clone()
    }

    fn clone_into(&self, target: &mut T) {
        target.clone_from(self);
    }
}
```

 I prefer to use `clone()` where it is possible in order not to complicate things.

2. Applied.

Reply:
> Maybe we'll change the signature of our parser to Fn(Vec<Token>) -> IResult<Vec<Token>, Token>

It reasonable in theory, but this will forces us to memmove all vector when we cut first element. It is cheaper to clone `Token`.